content/courses/advanced-ada/chapters/subprograms.rst

```diff
@@ -8,20 +8,631 @@ Subprograms
+
+    package Operations is
+
+       procedure Add (Left  : in out Integer);
+
+       procedure Add (Left  : in out Integer;
+                      Right :        Float := 1.0);
+
+    end Operations;
+
+    package body Operations is
+
+       procedure Add (Left  : in out Integer) is
+       begin
+          Left := Left + 1;
+       end Add;
+
+       procedure Add (Left  : in out Integer;
+                      Right :        Float := 1.0) is
+       begin
+          Left := Left + Integer (Right);
+       end Add;
+
+    end Operations;
+
+    with Operations; use Operations;
+
+    procedure Show_Param_Association is
+       A : Integer := 5;
+    begin
+       Add (A);
+       --  ERROR: cannot decide which
+       --         procedure to take
+    end Show_Param_Association;
+
+As we see in this example, the :ada:`Add` procedure is overloaded. The first
+instance has one parameter, and the second instance has two parameters, where
+the second parameter has a default value. When we call :ada:`Add` with just one
+parameter, the compiler cannot decide whether we intend to call
+
+- the first instance of :ada:`Add` with one parameter
+
+or
+
+- the second instance of :ada:`Add` using the default value for the second
+  parameter.
+
+In this specific case, there are multiple options to solve the issue, but all of
+them involve redesigning the package specification:
+
+- we could just rename one of :ada:`Add` procedures (thereby eliminating the
+  subprogram overloading);
+
+- we could rename the first parameter of one of the :ada:`Add` procedures and
+  use named parameter association in the call to the procedure;
+
+   - For example, we could rename the parameter to :ada:`Value` and call
+     :ada:`Add (Value => A)`.
+
+- remove the default value from the second parameter of the second instance of
+  :ada:`Add`.
+
+
+Overlapping actual parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When we have more than one :ada:`out` or :ada:`in out` parameters in a
+subprogram, we might run into the situation where the actual parameter overlaps
+with another parameter. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Subprograms.Illegal_Calls
+    :class: ada-expect-compile-error
+
+    package Machine_X is
+
+       procedure Update_Value (V1 : in out Integer;
+                               V2 :    out Integer);
+
+    end Machine_X;
+
+    package body Machine_X is
+
+       procedure Update_Value (V1 : in out Integer;
+                               V2 :    out Integer) is
+       begin
+          V1 := V1 + 1;
+          V2 := V2 + 1;
+       end Update_Value;
+
+    end Machine_X;
+
+    with Machine_X; use Machine_X;
+
+    procedure Show_By_Copy_By_Ref_Params is
+       A : Integer := 5;
+    begin
+       Update_Value (A, A);
+    end Show_By_Copy_By_Ref_Params;
+
+In this case, we're using :ada:`A` for for output parameters in the call to
+:ada:`Update_Value`. Having one variable in more than one output parameter is
```

Comment:
"Having one variable in more than one output parameter is" is a little awkward. Maybe "Passing one variable to more than one output parameter in a given call is" or something like that.

Reply:
Yes, your suggestion is much better.